File: backend/src/main/java/xyz/stasiak/recipai/shoppinglists/ShoppingListItem.java
package xyz.stasiak.recipai.shoppinglists;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Objects;
import java.util.UUID;

@Entity
@Table(name = "shopping_list_items")
@Getter
@Setter
@ToString
@NoArgsConstructor
class ShoppingListItem {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "shopping_list_id", nullable = false)
    private UUID shoppingListId;

    @Column(nullable = false)
    private String name;

    @Column(precision = 12, scale = 3)
    private BigDecimal quantity;

    @Column(length = 64)
    private String unit;

    @Column(nullable = false)
    private Boolean checked = false;

    @Column(precision = 21, scale = 12, nullable = false)
    private BigDecimal position;

    @Version
    @Column(nullable = false)
    private Long version;

    ShoppingListItem(UUID shoppingListId, String name, BigDecimal quantity, String unit, boolean checked, BigDecimal position) {
        this.shoppingListId = shoppingListId;
        this.name = name;
        this.quantity = quantity;
        this.unit = unit;
        this.position = normalizePositionScale(position);
        this.checked = checked;
    }

    void check() {
        this.checked = true;
    }

    void uncheck() {
        this.checked = false;
    }

    void setPosition(BigDecimal position) {
        this.position = normalizePositionScale(position);
    }

    private static BigDecimal normalizePositionScale(BigDecimal position) {
        return position != null ? position.setScale(12, RoundingMode.HALF_UP) : null;
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        ShoppingListItem that = (ShoppingListItem) o;
        return Objects.equals(id, that.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }
}

File: mobile/lib/features/shopping_list/shopping_list_item.dart
class ShoppingListItem {
  final String id;
  final String name;
  final double? quantity;
  final String? unit;
  final bool checked;
  final int position;

  const ShoppingListItem({
    required this.id,
    required this.name,
    required this.quantity,
    required this.unit,
    required this.checked,
    required this.position,
  });

  factory ShoppingListItem.fromJson(Map<String, dynamic> json) {
    return ShoppingListItem(
      id: json['id'] as String,
      name: json['name'] as String,
      quantity: json['quantity'] as double?,
      unit: json['unit'] as String?,
      checked: json['checked'] as bool,
      position: json['position'] as int,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'quantity': quantity,
      'unit': unit,
      'checked': checked,
      'position': position,
    };
  }
}
